File: weekly-contest/weekly-contest-190/3-pseudo-palindromic-paths-in-a-binary-tree/bit.js
/**
 * DFS + 位压缩
 */
var pseudoPalindromicPaths  = function(root) {
  let res = 0;

  function dfs(node, path) {
    path ^= 1 << node.val;
    
    if (!node.left && !node.right) {
      if (isPseudoPalindromic(path)) {
        ++res;
      }
    } else {
      if (node.left) dfs(node.left, path);
      if (node.right) dfs(node.right, path);
    }
  }

  dfs(root, 0);
  return res;
};

/**
 * 判断是否伪回文
 * 偶回文：全0
 * 奇回文：只有1个1
 * 能统一为一种情况
 */
function isPseudoPalindromic(x) {
  return !(x & (x - 1));
}

File: weekly-contest/weekly-contest-190/3-pseudo-palindromic-paths-in-a-binary-tree/bit.cpp
class Solution {
public:
    int pseudoPalindromicPaths (TreeNode* root) {
      res = 0;
      dfs(root, 0);
      return res;
    }

    void dfs(TreeNode* node, int path) {
      path ^= 1 << node->val;
      
      if (!node->left && !node->right) {
        if (isPseudoPalindromic(path)) {
          ++res;
        }
      } else {
        if (node->left) dfs(node->left, path);
        if (node->right) dfs(node->right, path);
      }
    }

    bool isPseudoPalindromic(int x) {
      return !(x & (x - 1));
    }
private:
    int res;
};
